trunk/modules/python/dataset/reader.py

```diff
@@ -144,38 +205,76 @@ def makePILImageBandContiguous(img, log=None) :
         a = np.asarray(img.getdata(), dtype=t.config.floatX)
         return np.resize(statisticalNorm(a), (1, img.size[1], img.size[0]))
 
-def readPILImage(image, log=None) :
+def readPILImage(image, log=None, **kwargs) :
     '''This method should be used for all regular image formats from JPEG,
        PNG, TIFF, etc. A PIL error may originate from this method if the image
        format is unsupported.
     '''
     from PIL import Image
-    img = Image.open(image)
-    img.load() # because PIL can be lazy
-    return makePILImageBandContiguous(img)
+    from io import BytesIO
+
+    # PIL images can be a bunch of different types
+    # with different memory layouts that are nasty
+    # to recreate (putting them as numpy arrays
+    # disturbs their delicate nature)
+    #
+    # So for 'raw' input, just punt on interpreting
+    # them at all and just grab the raw file bytes.
+    # 
+    # When we need the 'processed' version, PIL
+    # is smart enough to decode these in memory
 
-def readImage(image, log=None) :
+    if type(image) == tuple:
```

Comment:
Here always call makePILImageBandContiguous() however remove the call to statisticalNorm() on the data. This will always return the raw data cast into floatX and in the 3-channel format we need. It will also greatly simplify this logic.